WIP: Split deploy_node and prepare etcd members for cluster addition

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt'

**Context**: 

See #2028 

**Summary**:

- Split deploy_node into generic stages
- Add prepared state to prepare etcd deps and register etcd node into etcd cluster
- Add running state to create manifest and wait for member to be healthy


**Acceptance criteria**: 


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2028 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2028-ensure-etcd-is-running-before-registering-members` into
`development/2.4`
* [w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #a62df6e25135fc7693e0c4e2cdd06721de67d45c on the integration branch
`w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members` is merging a branch which is neither the current
branch `bugfix/2028-ensure-etcd-is-running-before-registering-members` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2028-ensure-etcd-is-running-before-registering-members` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck committed:
salt: Split node deployment procedure

This commit splits node deployment procedure into
a sequence of stages.

Initially, expanding the etcd cluster from 1 to 2 nodes puts
the cluster in a blocked state. This happens when a new member is declared
and not yet started.
If this time to start is too long, liveness probes on the initial etcd Pod will
fail, then it will be restarted and then enter a back-off loop.
Even after the new etcd Pod has been created, the new one won't be able to start,
since the initial is in back-off and will enter a back-off as well.

To prevent the above situation, we pre-pull the etcd image on the target Node,
which ensures the shortest time possible between the new member being registered
and its corresponding Pod being created. Note that this reordering is inspired by
kubespray

TeddyAndrieux commented on salt/metalk8s/kubernetes/etcd/installed.sls:
```diff
@@ -80,3 +80,24 @@ Create local etcd Pod manifest:
+    - status: 200
+    - match: '{"health":"true"}'
+    - require:
+      - module: Delay after etcd pod is deployed
```
If you do this here you can basically remove the `metalk8s.kubernetes.etcd.healthy` sls (and change it in `metalk8s.orchestrate.etcd`)

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +113,21 @@ Wait minion available:
+    - require:
+      - salt: Set grains
+      - salt: Refresh the mine
+      - salt: Refresh and check pillar before highstate
```
Add a `require_in` here on `Run the highstate`

TeddyAndrieux commented on salt/metalk8s/roles/etcd/prepared.sls:
```diff
@@ -0,0 +1,7 @@
+    - name: state.orchestrate
+    - pillar: {{ pillar | json  }}
+    - mods:
+      - metalk8s.orchestrate.register_etcd
```
Maybe an `include` of this state is enough here.

Also it's weird to me to have a sls in `roles` directory that need to run on the salt-master (cc @NicolasT )

TeddyAndrieux commented on salt/metalk8s/roles/etcd/running.sls:
```diff
@@ -0,0 +1,3 @@
+# creates the manifest YAML and wait for the member to be up and running
+include:
+  - metalk8s.kubernetes.etcd
```
This state really needed ?

For me it's not needed and we shouldn't create a state like that except if we want somewhere to run only this state and not all the `roles.etcd` (which is not the case for the moment)

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +113,21 @@ Wait minion available:
+Prepare etcd dependencies:
+  salt.state:
+    - tgt: {{ node_name }}
+    - sls: metalk8s.roles.etcd.prepared
```
You can't run this state on the minion as this is something that need `etcd3` python lib which is only available (by default) on the salt-master

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +113,23 @@ Wait minion available:
+  salt.runner:
+    - tgt: {{ node_name }}
+    - sls: metalk8s.kubernetes.etcd.prepared
+    - pillar: {{ pillar | json }}
```
That cannot work ! 
The `metalk8s.orchestrate.register_etcd` need to run on the salt-master not on the minion so you cannot use `salt.runner` targetting a node.

IMO you should call the register here and remove all `prepared` stuff as the prepare is only registering and cannot run on the salt-minion

gdemonet commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +113,35 @@ Wait minion available:
+      - salt: Refresh and check pillar before highstate
+      - salt: Prepare etcd dependencies
+    - require_in:
+      - salt: Run the highstate
```
Maybe add `metalk8s.kubernetes.etcd` between registering and highstate, to make sure `etcd` is unavailable the least amount of time possible. WDYT @NicolasT @TeddyAndrieux ?

TeddyAndrieux replied:
I think/hope that etcd cluster work when a node get registered but not joined. So IMO no.

gdemonet replied:
> etcd cluster work when a node get registered but not joined

No it does not

NicolasT replied:
When going from 1 to 2 nodes indeed it does not, so we want to make the "time" during which 2 nodes are registered but only 0 or 1 of them are running as short as possible, i.e., apply an as-specific-state as possible.

TeddyAndrieux replied:
Then I suggest we use the `skip_roles` so, we first run highstate without `etcd` and then we do etcd steps in the right order

gdemonet reviewed (commented):
Other than a suggestion to deploy `etcd` manifest before running highstate, looks good to me!

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +113,35 @@ Wait minion available:
 
 {%- endif %}
 
+{%- if 'etcd' in pillar.get('metalk8s', {}).get('nodes', {}).get(node_name, {}).get('roles', []) %}
```
Please skip this part if `etcd` is in `skip_roles`

TeddyAndrieux commented on salt/metalk8s/roles/etcd/prepared.sls:
```diff
@@ -0,0 +1,4 @@
+include:
+  - metalk8s.roles.internal.node-without-calico
+  - metalk8s.container-engine.containerd
```
Hmmm it's already part of `metalk8s.roles.internal.node-without-calico` so not needed

TeddyAndrieux commented on salt/metalk8s/kubernetes/etcd/prepared.sls:
```diff
@@ -0,0 +1,9 @@
+  containerd.image_managed:
+    - name: {{ image }}
+    - require:
+      - metalk8s_package_manager: Install and configure cri-tools
```
Please no ! Remove this require

gdemonet commented:
Another thing, the commit message explains well *what* has been done (splitting `deploy_node` and handling `etcd` first), but it doesn't explain *why*, which is crucial in this case.

First, explain the issue at hand: expanding the `etcd` cluster from 1 to 2 necessarily puts the cluster in a blocked state during the time the new member is declared and not yet started. If this time is too long, liveness probes on the initial `etcd` Pod will fail, it will be restarted, entering a back-off loop. Even after the new `etcd` Pod has been created, this one won't be able to start, since the initial is in back-off, and will enter a back-off as well.

Then, explain that this re-ordering, inspired by what kubespray does for pre-pulling the `etcd` image on the target Node, ensures the shortest time possible between the new member being registered and its corresponding Pod being created.

TeddyAndrieux commented on salt/metalk8s/kubernetes/etcd/prepared.sls:
```diff
@@ -0,0 +1,9 @@
+{%- from "metalk8s/repo/macro.sls" import build_image_name with context %}
+
+{%- set image = build_image_name("etcd") -%}
```
Do we really need to set this jinja var ? Putting directly `build_image_name("etcd")` in `name` of `containerd.image_managed` is simpler IMO

NicolasT commented on salt/metalk8s/roles/etcd/prepared.sls:
```diff
@@ -0,0 +1,3 @@
+include:
```
This is not a 'role', not sure it belongs here.

Ebaneck replied:
Done

NicolasT commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +114,35 @@ Wait minion available:
+  salt.state:
+    - tgt: {{ node_name }}
+    - saltenv: metalk8s-{{ version }}
+    - sls: metalk8s.roles.etcd.prepared
```
See other comment.

Ebaneck replied:
Done

TeddyAndrieux commented on salt/metalk8s/roles/etcd/init.sls:
```diff
@@ -1,3 +1,3 @@
 include:
-  - metalk8s.roles.internal.node-without-calico
-  - metalk8s.kubernetes.etcd
+  - metalk8s.kubernetes.etcd.configured
```
This state does not exist and IMO you should not change the roles state but only include the `prepared` in the `metalk8s.kubernetes.etcd.init`or directly in `metalk8s.kubernetes.etcd.installed`

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -1,7 +1,12 @@
+{%- set roles_to_skip = pillar.metalk8s.nodes[node_name].get('skip_roles', []) %}
+
+{#- Add etcd to skip_roles in order not to apply all the highstate #}
+{%- set skip_roles = roles_to_skip.append('etcd') %}
```
`append` do not return the list it's just alter the list so you can't do this as `skip_roles` will always be empty

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +115,42 @@ Wait minion available:
+      - metalk8s.roles.internal.node-without-calico
+      - metalk8s.kubernetes.etcd.prepared
+    - require_in:
+      - salt: Run the highstate
```
Remove this one only keep the `require_in` of the last one

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +115,42 @@ Wait minion available:
+      - salt: Prepare etcd dependencies
+    - require_in:
+      - salt: Run the highstate
+      - salt: Install and wait for healthy etcd member
```
ditto, and you don't need to `require_in` on `Install and wait for healthy etcd member` since you already `require` it in next state

TeddyAndrieux commented on salt/metalk8s/orchestrate/deploy_node.sls:
```diff
@@ -113,6 +115,47 @@ Wait minion available:
+    - timeout: 120
+    - sls: metalk8s.kubernetes.etcd.installed
+    - require:
+      - salt: Register the node into etcd cluster
```
Since register break the etcd cluster from 1 to 2 IMO no point to register before installing etcd.
I suggest we register just after `etcd.installed` then we shouldn't get the error we have before as `metalk8s.kubernetes.etcd.installed` will basically only render one file

TeddyAndrieux committed:
salt: Do etcd healthcheck by default and add pillar value to skip

When we run `etcd` state we want to check `etcd` health by default but
when we deploy a new `etcd` the new member is not yet registered in the
`etcd` cluster so we need to skip the healthcheck of the `etcd` when
deploying this new `etcd` node, so adding a pillar value to skip
`etcd` healthcheck in the salt state

TeddyAndrieux committed:
salt: Add delay after `apiserver-proxy` manifest changes

Add a small delay right after `apiserver-proxy` manifest get changes so
that we didn't check the "old" apiserver-proxy

TeddyAndrieux committed:
salt: Add pillar value to skip `apiserver-proxy` healthcheck

If we try to deploy `apiserver-proxy` on an "master" node we will have
`weight=100` for the local apiserver but this local apiserver may not be
ready or configured yet, so we need to be able to skip the
`apiserver-proxy` healthcheck from the salt state

TeddyAndrieux committed:
salt: Add some refresh and check pillar in deploy node

During deploy node we restart some mandatory pod (salt-master,
repositories) at several place so we need to check pillar value
availability several time

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2028-ensure-etcd-is-running-before-registering-members` into
`development/2.4`
* [w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #edb8a3da12b20f4a2242b2af1d109192f41be898 on the integration branch
`w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members` is merging a branch which is neither the current
branch `bugfix/2028-ensure-etcd-is-running-before-registering-members` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2028-ensure-etcd-is-running-before-registering-members` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2028-ensure-etcd-is-running-before-registering-members` into
`development/2.4`
* [w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

Ebaneck commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/7399097303b91c5a3eedc2de3310d51e79b440ee) did not succeed in branch w/2.5/bugfix/2028-ensure-etcd-is-running-before-registering-members.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
